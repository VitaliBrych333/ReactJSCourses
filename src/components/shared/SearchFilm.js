import React, { useState, createRef } from 'react';
import { InputGroup, FormControl, Button } from 'react-bootstrap';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import AddPage from '../AddPage';
import useLocalStorageState from './useLocalStorageState';
import { fetchMovies } from '../../redux/actions/moviesActions';
import { showAddPage } from '../../redux/actions/windowActions';
import { setGenre } from '../../redux/actions/criteriaActions';

const StyledGroup = styled(InputGroup)`
  padding: 0 50px;
  margin: 0 auto;
  max-width: 700px;

  .input-group-append {
    background-color: bisque;
  }
`;

const StyleDiv = styled.div`
  a {
    color: #f65261;
  }

  a:hover {
    text-decoration: none;
  }

  button:hover > a {
    color: #fff;
  }
`;
const SearchFilm = (props) => {
  const { dispatch, sort, isShowAddPage } = props;

  const [controlValue, setValue] = useState({
    disabled: true,
    myRef: createRef(),
  });

  const [defaultValue, setState] = useLocalStorageState(
    'my-app-defaultValueSearch',
    ''
  );

  const handleClick = () => {
    dispatch(setGenre('All'));
    dispatch(fetchMovies(sort, controlValue.myRef.value));
  };

  const handleChange = () => {
    setValue({ disabled: !controlValue.myRef.value });
    setState(controlValue.myRef.value);
  };

  const handleAdd = () => {
    dispatch(showAddPage(true));
  };

  return (
    <>
      <StyleDiv>
        {isShowAddPage && <AddPage />}
        <Button
          className="add-movie"
          variant="outline-danger"
          onClick={handleAdd}
        >
          + Add movie
        </Button>
        <h1>Find your movie</h1>
      </StyleDiv>
      <StyledGroup className="mb-3">
        <FormControl
          placeholder="Please write the film name"
          ref={(controlInput) => (controlValue.myRef = controlInput)}
          onChange={handleChange}
          defaultValue={defaultValue}
        />
        <InputGroup.Append>
          <Button
            variant="outline-danger"
            onClick={handleClick}
            disabled={controlValue.disabled && !defaultValue}
          >
            Search
          </Button>
        </InputGroup.Append>
      </StyledGroup>
    </>
  );
};

SearchFilm.propTypes = {
  dispatch: PropTypes.func,
  sort: PropTypes.string,
  isShowAddPage: PropTypes.bool,
};

const mapStateToProps = (state) => ({
  sort: state.criteriaReducer.sort,
  isShowAddPage: state.windowReducer.isShowAddPage,
});

export default connect(mapStateToProps)(SearchFilm);
