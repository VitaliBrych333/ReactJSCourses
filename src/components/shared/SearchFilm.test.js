import React, { createRef } from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { FormControl, Button } from 'react-bootstrap';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import thunk from 'redux-thunk';
import SearchFilm from './SearchFilm';

Enzyme.configure({ adapter: new Adapter() });

describe('<SearchFilm/>', () => {
  const initialState = {
    movieReducer: {
      sort: 'sort',
    },
    windowReducer: {
      isShowAddPage: false,
    },
  };
  const middlewares = [thunk];
  const mockStore = configureStore(middlewares);
  let store;
  let wrapper;

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    store = mockStore(initialState);
    wrapper = shallow(
      <Provider store={store}>
        <SearchFilm />
      </Provider>
    );
  });

  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  it('render component', () => {
    expect(wrapper.find(SearchFilm).length).toEqual(1);
  });

  it('should equals to snapshot of SearchFilm when isShowAddPage is false', () => {
    const renderedValue = renderer
      .create(
        <Router>
          <Provider store={store}>
            <SearchFilm />
          </Provider>
        </Router>
      )
      .toJSON();
    expect(renderedValue).toMatchSnapshot();
  });

  it('should equals to snapshot of SearchFilm isShowAddPage is true', () => {
    const newInitialState = {
      movieReducer: {
        sort: 'sort',
      },
      windowReducer: {
        isShowAddPage: true,
      },
    };
    store = mockStore(newInitialState);
    const renderedValue = renderer
      .create(
        <Router>
          <Provider store={store}>
            <SearchFilm />
          </Provider>
        </Router>
      )
      .toJSON();
    expect(renderedValue).toMatchSnapshot();
  });

  it('should not call onChange and onClick', () => {
    const onChange = jest.fn();
    const onClick = jest.fn();
    const wrap = mount(
      <Router>
        <Provider store={store}>
          <SearchFilm />
        </Provider>
      </Router>
    );

    wrap.find(FormControl).at(0).simulate('change');
    wrap.find(Button).at(0).simulate('click');
    expect(onChange).toHaveBeenCalledTimes(0);
    expect(onClick).toHaveBeenCalledTimes(0);
  });

  it('should render button with string Search', () => {
    const realUseState = React.useState;
    const stubInitialState = {
      disabled: false,
      myRef: createRef(),
    };
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => realUseState(stubInitialState));
    render(
      <Router>
        <Provider store={store}>
          <SearchFilm />
        </Provider>
      </Router>
    );

    const countValue = screen.queryByText('Search');
    userEvent.click(countValue);

    expect(countValue.textContent).toBe('Search');
  });
});
