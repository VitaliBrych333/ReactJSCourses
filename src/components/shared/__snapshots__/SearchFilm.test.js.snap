// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchFilm/> should equals to snapshot of SearchFilm isShowAddPage is true 1`] = `
Array [
  <div
    className="sc-EHOje dACVym"
  >
    <section
      className="sc-bxivhb bjzxkO"
    >
      <div
        className="modal"
      >
        <section
          className="modal-main"
        >
          <div
            className="sc-htpNat iQwqzL"
          >
            <form
              className=""
            >
              <div
                className="sc-bwzfXH kbQCzJ"
              >
                <button
                  className="close"
                  onClick={[Function]}
                  type="button"
                >
                  <span>
                    Ã—
                  </span>
                </button>
                <h2>
                  Add movie
                </h2>
              </div>
              <div
                className="Add movie form-group"
              >
                <label
                  className="form-label"
                >
                  Title
                </label>
                <input
                  className="form-control"
                  name="title"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Title"
                  type="title"
                  value=""
                />
                <label
                  className="form-label"
                >
                  Release date
                </label>
                <input
                  className="form-control"
                  name="date"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Select Date"
                  type="date"
                  value=""
                />
                <label
                  className="form-label"
                >
                  Movie URL
                </label>
                <input
                  className="form-control"
                  name="url"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Movie URL here"
                  type="url"
                  value=""
                />
                <label
                  className="form-label"
                >
                  Example select
                </label>
                <div
                  className="select css-2b097c-container"
                  onKeyDown={[Function]}
                >
                  <div
                    className=" css-yk16xz-control"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <div
                      className=" css-g1d714-ValueContainer"
                    >
                      <div
                        className=" css-1wa3eu0-placeholder"
                      >
                        Select Genre
                      </div>
                      <div
                        className="css-b8ldur-Input"
                      >
                        <div
                          className=""
                          style={
                            Object {
                              "display": "inline-block",
                            }
                          }
                        >
                          <input
                            aria-autocomplete="list"
                            autoCapitalize="none"
                            autoComplete="off"
                            autoCorrect="off"
                            disabled={false}
                            id="react-select-2-input"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            spellCheck="false"
                            style={
                              Object {
                                "background": 0,
                                "border": 0,
                                "boxSizing": "content-box",
                                "color": "inherit",
                                "fontSize": "inherit",
                                "label": "input",
                                "opacity": 1,
                                "outline": 0,
                                "padding": 0,
                                "width": "1px",
                              }
                            }
                            tabIndex="0"
                            type="text"
                            value=""
                          />
                          <div
                            style={
                              Object {
                                "height": 0,
                                "left": 0,
                                "overflow": "scroll",
                                "position": "absolute",
                                "top": 0,
                                "visibility": "hidden",
                                "whiteSpace": "pre",
                              }
                            }
                          >
                            
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      className=" css-1hb7zxy-IndicatorsContainer"
                    >
                      <span
                        className=" css-1okebmr-indicatorSeparator"
                      />
                      <div
                        aria-hidden="true"
                        className=" css-tlfecz-indicatorContainer"
                        onMouseDown={[Function]}
                        onTouchEnd={[Function]}
                      >
                        <svg
                          aria-hidden="true"
                          className="css-6q0nyr-Svg"
                          focusable="false"
                          height={20}
                          viewBox="0 0 20 20"
                          width={20}
                        >
                          <path
                            d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
                <label
                  className="form-label"
                >
                  Overview
                </label>
                <input
                  className="form-control"
                  name="overview"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Overview here"
                  type="text"
                  value=""
                />
                <label
                  className="form-label"
                >
                  Runtime
                </label>
                <input
                  className="form-control"
                  name="runtime"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Runtime here"
                  type="number"
                  value=""
                />
              </div>
              <div
                className="sc-bdVaJa kYETlp"
              >
                <button
                  className="btn btn-primary"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Reset
                </button>
                <button
                  className="btn btn-primary"
                  disabled={true}
                  onClick={[Function]}
                  type="button"
                >
                  Submit
                </button>
              </div>
            </form>
          </div>
        </section>
      </div>
    </section>
    <button
      className="add-movie btn btn-outline-danger"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      + Add movie
    </button>
    <h1>
      Find your movie
    </h1>
  </div>,
  <div
    className="sc-ifAKCX hOgznM mb-3 input-group"
  >
    <input
      className="form-control"
      defaultValue=""
      onChange={[Function]}
      placeholder="Please write the film name"
    />
    <div
      className="input-group-append"
    >
      <button
        className="btn btn-outline-danger"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Search
      </button>
    </div>
  </div>,
]
`;

exports[`<SearchFilm/> should equals to snapshot of SearchFilm when isShowAddPage is false 1`] = `
Array [
  <div
    className="sc-EHOje dACVym"
  >
    <button
      className="add-movie btn btn-outline-danger"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      + Add movie
    </button>
    <h1>
      Find your movie
    </h1>
  </div>,
  <div
    className="sc-ifAKCX hOgznM mb-3 input-group"
  >
    <input
      className="form-control"
      defaultValue=""
      onChange={[Function]}
      placeholder="Please write the film name"
    />
    <div
      className="input-group-append"
    >
      <button
        className="btn btn-outline-danger"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        Search
      </button>
    </div>
  </div>,
]
`;
